lib/puppet/transaction.rb

```diff
@@ -358,15 +358,16 @@ def prefetch(provider_class, resources)
     Puppet.debug "Prefetching #{provider_class.name} resources for #{type_name}"
     begin
       provider_class.prefetch(resources)
-    rescue Exception => detail
+    rescue LoadError, StandardError => detail
+      #TRANSLATORS `prefetch` is a function name and should not be translated
+      message = _("Could not prefetch %{type_name} provider '%{name}': %{detail}") % { type_name: type_name, name: provider_class.name, detail: detail }
+      Puppet.log_exception(detail, message)
+
       if !detail.is_a?(LoadError) && !detail.is_a?(Puppet::MissingCommand)
         raise unless Puppet.settings[:future_features]
-        
+
```

Comment:
There's some subtly for exceptions that derive from `StandardError` that aren't `MissingCommand`. I think it'd be better to break it out into different rescue clauses, from most specific to least. In order to preserve the current 5.x/6.x behavior I'd do:

```ruby
rescue LoadError, Puppet::MissingCommand => e
  log_message
  @prefetch_failed_providers[type_name][provider_class.name] = true
rescue StandardError => e
  log_message
  raise unless Puppet.settings[:future_features]
  @prefetch_failed_providers[type_name][provider_class.name] = true
end
```

In Puppet 7, I think we should drop the `future_features` logic, and collapse the clauses. I think it would also be good to rescue `ScriptError` instead of just `LoadError`.

```ruby
rescue ScriptError, StandardError => e
  Puppet.log_exception(...)
  @prefetch_failed_providers[type_name][provider_class.name] = true
end
```

We really don't want to rescue `Exception`, because it could be `SystemExit`, `Interrupt`, `NoMemoryError`, etc. and that can interfere with Ruby itself, e.g. Ctrl-C processing.